cfme/common/provider.py

```diff
@@ -153,47 +150,153 @@ def create(self, cancel=False, validate_credentials=True, check_existing=False,
+                logger.info('Setting up provider: %s', self.key)
+                navigate_to(self, 'Add')
+                fill(self.properties_form, self._form_mapping(True, hawkular=False,
+                                                              **self.__dict__))
+                for cred in self.credentials:
+                    fill(self.credentials[cred].form, self.credentials[cred],
+                         validate=validate_credentials)
+                self._submit(cancel, self.add_provider_button)
+                if not cancel:
+                    flash.assert_message_match('{} Providers '
+                                               '"{}" was saved'.format(self.string_name, self.name))
+            if validate_inventory:
+                self.validate()
+            return created
+
+    def update(self, updates, endpoints=None, cancel=False, validate_credentials=True):
         """
         Updates a provider in the UI.  Better to use utils.update.update context
         manager than call this directly.
 
         Args:
            updates (dict): fields that are changing.
+           endpoints (list of Endpoints): objects which hold information about provider's endpoints
            cancel (boolean): whether to cancel out of the update.
+           validate_credentials (boolean): whether credentials have to be validated
         """
-        navigate_to(self, 'Edit')
-        fill(self.properties_form, self._form_mapping(**updates))
-        for cred in self.credentials:
-            fill(self.credentials[cred].form, updates.get('credentials', {}).get(cred, None),
-                 validate=validate_credentials)
-        self._submit(cancel, self.save_button)
-        name = updates.get('name', self.name)
-        if not cancel:
-            if BZ.bugzilla.get_bug(1436341).is_opened and version.current_version() > '5.8':
-                logger.warning('Skipping flash message verification because of BZ 1436341')
-                return
-            flash.assert_message_match(
-                '{} Provider "{}" was saved'.format(self.string_name, name))
+        from cfme.infrastructure.provider import InfraProvider
+        if self.one_of(InfraProvider):
+            main_view = navigate_to(self, 'Edit')
+            # todo: to replace/merge this code with create
+            # update values:
+            # filling main part of dialog
+            if updates:
+                main_view.fill(updates)
+
+            # filling endpoints
+            if endpoints:
+                endpoints = endpoints if isinstance(endpoints, list) else [endpoints, ]
+
+                for endpoint in endpoints:
+                    endp_view = getattr(self.endpoints_form(parent=main_view), endpoint.name)
+                    endp_view.fill(endpoint.view_value_mapping)
+
+                    # filling credentials
+                    cur_endpoint = next(
+                        endp for endp in self.endpoints if endp.name == endpoint.name)
```

Comment:
**Optional**
Comment here describing you're comparing the enpoint passed to the object's endpoint